test/functional/test_framework/test_framework.py

```diff
@@ -202,51 +175,91 @@ def main(self):
             logging.shutdown()
             sys.exit(TEST_EXIT_FAILED)
 
-    # Public helper methods. These can be accessed by the subclass test scripts.
+    # Methods to override in subclass test scripts.
+    def set_test_params(self):
+        """Override this method to change default values for number of nodes, topology, etc"""
+        pass
 
-    def start_node(self, i, dirname, extra_args=None, rpchost=None, timewait=None, binary=None, stderr=None):
-        """Start a bitcoind and return RPC connection to it"""
+    def add_options(self, parser):
+        """Override this method to add command-line options to the test"""
+        pass
 
-        if extra_args is None:
-            extra_args = []
-        if binary is None:
-            binary = os.getenv("BITCOIND", "bitcoind")
-        node = TestNode(i, dirname, extra_args, rpchost, timewait, binary, stderr, self.mocktime, coverage_dir=self.options.coveragedir)
-        node.start()
-        node.wait_for_rpc_connection()
+    def setup_chain(self):
+        """Override this method to customize blockchain setup"""
+        self.log.info("Initializing test directory " + self.options.tmpdir)
+        if self.setup_clean_chain:
+            self._initialize_chain_clean(self.options.tmpdir, self.num_nodes)
+        else:
+            self._initialize_chain(self.options.tmpdir, self.num_nodes, self.options.cachedir)
 
-        if self.options.coveragedir is not None:
-            coverage.write_all_rpc_commands(self.options.coveragedir, node.rpc)
+    def setup_network(self):
+        """Override this method to customize test network topology"""
+        self.setup_nodes()
+
+        # Connect the nodes as a "chain".  This allows us
+        # to split the network between nodes 1 and 2 to get
+        # two halves that can work on competing chains.
+        for i in range(self.num_nodes - 1):
+            connect_nodes_bi(self.nodes, i, i + 1)
+        self.sync_all()
+
+    def setup_nodes(self):
+        """Override this method to customize test node setup"""
+        extra_args = None
+        if hasattr(self, "extra_args"):
+            extra_args = self.extra_args
+        self.add_nodes(self.num_nodes, self.options.tmpdir, extra_args)
+        self.start_nodes()
 
-        return node
+    def run_test(self):
+        """Override this method to define test logic"""
+        raise NotImplementedError
 
-    def start_nodes(self, num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
-        """Start multiple bitcoinds, return RPC connections to them"""
+    # Public helper methods. These can be accessed by the subclass test scripts.
+
+    def add_nodes(self, num_nodes, dirname, extra_args=None, rpchost=None, timewait=None, binary=None):
```

Comment:
Since `self` knows about `self.options.tmpdir`, and all calls to `self.add_nodes` passes `self.options.tmpdir` as `dirname` argument, it seems clean to just give it a default `None` and to set to `self.options.tmpdir if None` in `add_nodes`.
Or perhaps remove it entirely and always use `self.options.tmpdir`. Will it ever be anything else?